pkg/controller/connectivitycheck/connectivity_check_controller.go

```diff
@@ -0,0 +1,403 @@
+		kubeAPIServerEndpointsLister:      kubeInformersForNamespaces.InformersFor("openshift-kube-apiserver").Core().V1().Endpoints().Lister(),
+		kubeAPIServerServiceLister:        kubeInformersForNamespaces.InformersFor("openshift-kube-apiserver").Core().V1().Services().Lister(),
+		openshiftAPIServerEndpointsLister: kubeInformersForNamespaces.InformersFor("openshift-apiserver").Core().V1().Endpoints().Lister(),
+		openshiftAPIServerServiceLister:   kubeInformersForNamespaces.InformersFor("openshift-apiserver").Core().V1().Services().Lister(),
+		infrastructureLister:              configInformers.Config().V1().Infrastructures().Lister(),
+	}
+	return c.WithPodNetworkConnectivityCheckFn(generator.generate)
+}
+
+type networkConnectivityCheckController struct {
+	connectivitycheckcontroller.ConnectivityCheckController
+}
+
+type connectivityCheckTemplateProvider struct {
+	operatorClient                    v1helpers.OperatorClient
+	operatorcontrolplaneClient        *operatorcontrolplaneclient.Clientset
+	diagnosticsPodLister              corev1listers.PodLister
+	diagnosticsEndpointsLister        corev1listers.EndpointsLister
+	diagnosticsServiceLister          corev1listers.ServiceLister
+	kubeAPIServerEndpointsLister      corev1listers.EndpointsLister
+	kubeAPIServerServiceLister        corev1listers.ServiceLister
+	openshiftAPIServerEndpointsLister corev1listers.EndpointsLister
+	openshiftAPIServerServiceLister   corev1listers.ServiceLister
+	infrastructureLister              configv1listers.InfrastructureLister
+}
+
+func (c *connectivityCheckTemplateProvider) generate(ctx context.Context, syncContext factory.SyncContext) ([]*v1alpha1.PodNetworkConnectivityCheck, error) {
+	var templates []*v1alpha1.PodNetworkConnectivityCheck
+	// kas service IP
+	templates = append(templates, c.getTemplatesForKubernetesServiceMonitorService(syncContext.Recorder())...)
+	// kas default service IP
+	templates = append(templates, c.getTemplatesForKubernetesDefaultServiceCheck(syncContext.Recorder())...)
+	// each kas endpoint IP
+	templates = append(templates, c.getTemplatesForKubernetesServiceEndpointsChecks(syncContext.Recorder())...)
+	// oas service IP
+	templates = append(templates, c.getTemplatesForOpenShiftAPIServerServiceCheck(syncContext.Recorder())...)
+	// each oas endpoint IP
+	templates = append(templates, c.getTemplatesForOpenShiftAPIServerServiceEndpointsChecks(syncContext.Recorder())...)
+	// each api load balancer hostname
+	templates = append(templates, c.getTemplatesForAPILoadBalancerChecks(syncContext.Recorder())...)
+	// generic pod service IP
+	templates = append(templates, c.getTemplatesForGenericPodServiceCheck(syncContext.Recorder())...)
+	// each generic pod endpoint IP
+	templates = append(templates, c.getTemplatesForGenericPodServiceEndpointsChecks(syncContext.Recorder())...)
+
+	pods, err := c.diagnosticsPodLister.List(labels.Set{"app": "network-check-source"}.AsSelector())
```

Comment:
Are we sure this is consistently done before the render phase and that this call actually returns all pods?

Reply:
yeah, it will when they are deployed.
FWIW, there is only one network-check-source.

Reply:
I understand that there is only one network-check-source. However that will be deployed after the render phase is completed by the CNO, and I don't understand how we guarantee that this line is called _after_ that part so that `pods` is not empty?

Reply:
Ah ok, you said pods, that's why I pointed out it's only *one* pod.
Why it matters if they are deployed after?
This will reconcile as the pod come up, and the controller doesn't update the network checks till it it can perform the check between source and dest.

Reply:
If you like, you could create the pods in this controller, rather than in the network controller. That would be a reasonable logical division.